
package WSPROTO;

/* Client messages */

message ClientMessage {
	optional uint32 sequence = 1;
	optional LoginRequest loginrequest = 2;
	optional SelectRequest selectrequest = 3;
	optional FunctionRequest functionrequest = 4;
}

message LoginRequest {
	optional string login = 1;
	optional string password = 2;
}

message SelectRequest {
	required string table = 1;
}

message FunctionParam {
	optional string s = 1;
	optional int32 i = 2;
}

message FunctionRequest {
	required string name = 1;
	repeated FunctionParam params = 2;
}

/* Server messages */

message ServerMessage {
	optional uint32 sequence = 1;
	optional ErrorResponse error = 2;
	optional LoginResponse loginresponse = 3;
	optional SelectResponse selectresponse = 4;
}

enum ErrorCode {
	UNKNOWN = 0;
	SERVER_ERROR = 1;
	NOT_LOGGED_IN = 2;
};

message ErrorResponse {
	required ErrorCode code = 1 [default = UNKNOWN];
	optional string message = 2;
	optional bool fatal = 3;
}

message LoginResponse {
	required uint32 status = 1;
}

message SelectResponse {
	repeated string columns = 1;
	repeated string data = 2;
}
